File: android/app/src/main/java/my/ssdid/wallet/domain/transport/RegistryApi.kt
package my.ssdid.wallet.domain.transport

import my.ssdid.sdk.domain.model.DidDocument
import my.ssdid.wallet.domain.transport.dto.*
import retrofit2.http.*

interface RegistryApi {
    @GET("api/registry/info")
    suspend fun getRegistryInfo(): RegistryInfoResponse

    @POST("api/did")
    suspend fun registerDid(@Body request: RegisterDidRequest): RegisterDidResponse

    @GET("api/did/{did}")
    suspend fun resolveDid(@Path("did") did: String): DidDocument

    @PUT("api/did/{did}")
    suspend fun updateDid(@Path("did") did: String, @Body request: UpdateDidRequest): RegisterDidResponse

    @HTTP(method = "DELETE", path = "api/did/{did}", hasBody = true)
    suspend fun deactivateDid(@Path("did") did: String, @Body request: DeactivateDidRequest)

    @POST("api/did/{did}/challenge")
    suspend fun createChallenge(@Path("did") did: String): ChallengeResponse

    @POST("api/did/{did}/pair")
    suspend fun initPairing(@Path("did") did: String, @Body request: PairingInitRequest): PairingInitResponse

    @POST("api/did/{did}/pair/{pairingId}/join")
    suspend fun joinPairing(@Path("did") did: String, @Path("pairingId") pairingId: String, @Body request: PairingJoinRequest): PairingJoinResponse

    @GET("api/did/{did}/pair/{pairingId}")
    suspend fun getPairingStatus(@Path("did") did: String, @Path("pairingId") pairingId: String): PairingStatusResponse

    @POST("api/did/{did}/pair/{pairingId}/approve")
    suspend fun approvePairing(@Path("did") did: String, @Path("pairingId") pairingId: String, @Body request: PairingApproveRequest)
}

File: sdk/ios/SsdidCore/Sources/SsdidCore/Domain/Transport/RegistryApi.swift
import Foundation

/// Registry API client for DID operations and device pairing.
public final class RegistryApi: @unchecked Sendable {

    private let client: SsdidHttpClient
    private let baseURL: String

    public     init(client: SsdidHttpClient, baseURL: String) {
        self.client = client
        self.baseURL = baseURL
    }

    // MARK: - Registry Info

    public     func getRegistryInfo() async throws -> RegistryInfoResponse {
        return try await client.get(
            url: "\(baseURL)/api/registry/info",
            responseType: RegistryInfoResponse.self
        )
    }

    // MARK: - DID Operations

    public     func registerDid(request: RegisterDidRequest) async throws -> RegisterDidResponse {
        return try await client.post(
            url: "\(baseURL)/api/did",
            body: request,
            responseType: RegisterDidResponse.self
        )
    }

    public     func resolveDid(did: String) async throws -> DidDocument {
        let encodedDid = did.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? did
        return try await client.get(
            url: "\(baseURL)/api/did/\(encodedDid)",
            responseType: DidDocument.self
        )
    }

    public     func updateDid(did: String, request: UpdateDidRequest) async throws -> RegisterDidResponse {
        let encodedDid = did.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? did
        return try await client.request(
            method: "PUT",
            url: "\(baseURL)/api/did/\(encodedDid)",
            body: request,
            responseType: RegisterDidResponse.self
        )
    }

    public     func deactivateDid(did: String, request: DeactivateDidRequest) async throws {
        let encodedDid = did.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? did
        try await client.requestVoid(
            method: "DELETE",
            url: "\(baseURL)/api/did/\(encodedDid)",
            body: request
        )
    }

    public     func createChallenge(did: String) async throws -> ChallengeResponse {
        let encodedDid = did.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? did
        return try await client.post(
            url: "\(baseURL)/api/did/\(encodedDid)/challenge",
            body: Empty(),
            responseType: ChallengeResponse.self
        )
    }

    // MARK: - Device Pairing

    public     func initPairing(did: String, request: PairingInitRequest) async throws -> PairingInitResponse {
        let encodedDid = did.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? did
        return try await client.post(
            url: "\(baseURL)/api/did/\(encodedDid)/pair",
            body: request,
            responseType: PairingInitResponse.self
        )
    }

    public     func joinPairing(did: String, pairingId: String, request: PairingJoinRequest) async throws -> PairingJoinResponse {
        let encodedDid = did.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? did
        return try await client.post(
            url: "\(baseURL)/api/did/\(encodedDid)/pair/\(pairingId)/join",
            body: request,
            responseType: PairingJoinResponse.self
        )
    }

    public     func getPairingStatus(did: String, pairingId: String) async throws -> PairingStatusResponse {
        let encodedDid = did.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? did
        return try await client.get(
            url: "\(baseURL)/api/did/\(encodedDid)/pair/\(pairingId)",
            responseType: PairingStatusResponse.self
        )
    }

    public     func approvePairing(did: String, pairingId: String, request: PairingApproveRequest) async throws {
        let encodedDid = did.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? did
        try await client.requestVoid(
            method: "POST",
            url: "\(baseURL)/api/did/\(encodedDid)/pair/\(pairingId)/approve",
            body: request
        )
    }
}
